src/Storages/StorageMongoDB.cpp

```diff
@@ -55,257 +62,342 @@ StorageMongoDB::StorageMongoDB(
+            configuration.collection = checkAndGetLiteralArgument<String>(engine_args[1], "database");
+            configuration.uri =  std::make_unique<mongocxx::uri>(checkAndGetLiteralArgument<String>(engine_args[0], "host"));
         }
         else
-        {
-            Poco::SharedPtr<Poco::MongoDB::OpMsgMessage> insert_request = db.createOpMsgMessage(collection_name);
-            insert_request->setCommandName(Poco::MongoDB::OpMsgMessage::CMD_INSERT);
-            insert_request->documents() = std::move(documents);
-            connection->sendRequest(*insert_request);
-        }
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                                "Storage MongoDB requires 2 or from to 5 to 6 parameters: "
+                                "MongoDB('host:port', 'database', 'collection', 'user', 'password' [, 'options']) or MongoDB('uri', 'collection').");
     }
 
-private:
+    configuration.checkHosts(context);
 
-    void insertValueIntoMongoDB(
-        Poco::MongoDB::Document & document,
-        const std::string & name,
-        const IDataType & data_type,
-        const IColumn & column,
-        size_t idx)
-    {
-        WhichDataType which(data_type);
+    return configuration;
+}
 
-        if (which.isArray())
-        {
-            const ColumnArray & column_array = assert_cast<const ColumnArray &>(column);
-            const ColumnArray::Offsets & offsets = column_array.getOffsets();
+std::string mongoFuncName(const std::string & func)
+{
+    if (func == "equals")
+        return "$eq";
+    if (func == "notEquals")
+        return "$ne";
+    if (func == "greaterThan" || func == "greater")
+        return "$gt";
+    if (func == "lessThan" || func == "less")
+        return "$lt";
+    if (func == "greaterOrEquals")
+        return "$gte";
+    if (func == "lessOrEquals")
+        return "$lte";
+    if (func == "in")
+        return "$in";
+    if (func == "notIn")
+        return "$nin";
+    if (func == "lessThan")
+        return "$lt";
+    if (func == "and")
+        return "$and";
+    if (func == "or")
+        return "$or";
+
+    return "";
+}
 
-            size_t offset = offsets[idx - 1];
-            size_t next_offset = offsets[idx];
+template <typename OnError>
+std::optional<bsoncxx::document::value> StorageMongoDB::visitWhereFunction(const ContextPtr & context, const FunctionNode * func, OnError on_error)
+{
+    if (func->getArguments().getNodes().empty())
+        return {};
 
-            const IColumn & nested_column = column_array.getData();
+    if (const auto & column = func->getArguments().getNodes().at(0)->as<ColumnNode>())
+    {
+        // Skip unknown columns, which don't belong to the table.
+        const auto & table = column->getColumnSource()->as<TableNode>();
+        if (!table)
+            return {};
+
+        // Skip columns from other tables in JOIN queries.
+        if (table->getStorage()->getStorageID() != this->getStorageID())
```

Comment:
In case query is self-join, we cannot compare just storage ids.

Reply:
In case of JOIN I added condition, which will allow building filter only if column in the left side.